### PR TITLE
Merge 2.16 commit 8b34fef into baremetal

### PR DESCRIPTION
**Pull request by AndrzejKurek**
This PR updates the baremetal branch with changes done in the 2.16 branch, up to commit 8b34fef specifically. 
Detailed description of resolved conflicts (as in the commit description, plus disabling a test in ssl-opt.sh):

Conflicts:
**mbedtls.doxyfile -** PROJECT_NAME - mbed TLS v2.16.6 chosen.
**doc_mainpage.h** - mbed TLS v2.16.6 version chosen.
**hmac_drbg.h** - line 260, extended description chosen.
            - line 313, extended description chosen.
            - line 338, extended description chosen.
**version.h** - 2.16.6 chosen.
**CMakeLists.txt** - 2.16.6 chosen.
**test_suite_version.data** - 2.16.6 chosen.
**Makefile** - 141 - manual correction - baremetal version of C_SOURCE_FILES
                 with variables for directories plus 2.16.6 CTAGS addition.
**pkparse.c** - lines 846 onwards - the asn1_get_nonzero_mpi implementation chosen.
**ssl_tls.c** - line 5269 - edited manually, left the ret=0, because baremetal has
            a different behaviour since commit 87b5626, but added a debug
            message that's new in 2.16.6.    
**all.sh:**
- component_build_deprecated - chosen the refactored version from 2.16.6,
                               but with extra flags from baremetal.
- rest of the _no_xxx tests - merged make options to have PTHREAD=1 and
                              other changes from 2.16.6 (like -O1 instead of -O0).
- component_build_arm_none_eabi_gcc_no_64bit_multiplication - added 
                              TINYCRYPT_BUILD=0 to the 2.16.6 version of make.

**x509/req_app.c** - left baremetal log but with mbedtls_exit( 0 ) call.
**x509/crl_app.c** - left baremetal log but with mbedtls_exit( 0 ) call.
**x509/cert_app.c** - left baremetal log but with mbedtls_exit( 0 ) call.
**ssl/ssl_mail_client.c** - left baremetal log but with mbedtls_exit( 0 ) call.
**ssl/ssl_pthread_server.c** - left baremetal log but with mbedtls_exit( 0 ) call.
**ssl/ssl_fork_server.c** - left baremetal log but with mbedtls_exit( 0 ) call.
**ssl_client1.c** - line 54 - left baremetal log but with mbedtls_exit( 0 ) call.
**ssl_client2.c** - line 54 - left baremetal log but with mbedtls_exit( 0 ) call.
              - line 132 - new options of both branches added.
              - skip close notify handled as in 2.16.6, but with `ssl` instead of `&ssl`.
              - Merged the 2.16.6 usage split with additional baremetal usages.
              - Merged options from baremetal and 2.16.6.
**ssl_server.c** - left baremetal log but with mbedtls_exit( 0 ) call.
**ssl_server2.c** - Merged the 2.16.6 usage split with additional baremetal usages.
**config.pl** - fixed missing defines from the documentation, removed duplicates,
            and reorganised so that the documentation and excluded list
            are ordered in the same way.
**test_suite_x509parse.data** - only added the two new pathlen tests.
**x509_crt.c** - change the return code by removing
             MBEDTLS_ERR_X509_INVALID_EXTENSIONS, since it's added by
             x509_crt_frame_parse_ext not by an "or", but by "+=".
**Changelog** - Assigned all entries to appropriate sections.
**ssl-opt.sh** - line 8263 - merged options.
           - removed lines 1165 - 1176 - there was a duplicate test, probably
             an artifact of previous merges.
           - attacker-injected reconnection test - Applied https://github.com/ARMmbed/mbedtls/pull/3149.
**check-files.py** - sticked to old formatting.

### CODE REVIEW

**Comment by AndrzejKurek** [comment]
Added a couple of fixes for the failing tests. Also, test-ref-configs.pl failed on baremetal previously, so conditional compilation has been introduced into the internal x509 header to fix this issue.

The disabled test is fixed by PR https://github.com/ARMmbed/mbedtls/pull/3149, which should be adapted on top of this PR, preferably.

**Comment by AndrzejKurek** [comment]
The CI passes, and the only thing to do is either incorporate https://github.com/ARMmbed/mbedtls/pull/3149 into this PR, or merge it separately.

**Comment by AndrzejKurek** [comment]
Force pushed to incorporate Manuel's fix for the reconnection test.

File: library/ssl_tls.c [hunk 1]
```diff
@@ -1253,7 +1253,10 @@ int ssl_populate_transform( mbedtls_ssl_transform *transform,
                                    const unsigned char randbytes[64],
                                    int minor_ver,
                                    unsigned endpoint,
-                                   const mbedtls_ssl_context *ssl )
+#if !defined(MBEDTLS_SSL_HW_RECORD_ACCEL)
```
Comment (shelib01) [inline library/ssl_tls.c:1256]: @AndrzejKurek , what is the purpose of this flag MBEDTLS_SSL_HW_RECORD_ACCEL ?

File: library/ssl_tls.c [hunk 2]
```diff
@@ -5227,16 +5238,17 @@ static int ssl_check_dtls_clihlo_cookie(
  * that looks like a ClientHello.
  *
  * - if the input looks like a ClientHello without cookies,
- *   send back HelloVerifyRequest, then
- *   return MBEDTLS_ERR_SSL_HELLO_VERIFY_REQUIRED
+ *   send back HelloVerifyRequest, then return 0
```
Comment (shelib01) [inline library/ssl_tls.c:5241]: "send back HelloVerifyRequest" Was the behavior of previous 2.16.4/5 versions ?

File: scripts/assemble_changelog.py [hunk 3]
```diff
@@ -0,0 +1,505 @@
+#!/usr/bin/env python3
+
+"""Assemble Mbed TLS change log entries into the change log file.
+
```
Comment (shelib01) [inline scripts/assemble_changelog.py:4]: Is it completely new file in 2.16 or our branch didn't have it before for some reason ?

File: scripts/config.pl [hunk 4]
```diff
@@ -54,8 +56,13 @@
 #   MBEDTLS_AES_ONLY_128_BIT_KEY_LENGTH
 #   MBEDTLS_AES_ONLY_ENCRYPT
 #   MBEDTLS_AES_SCA_COUNTERMEASURES
+#   MBEDTLS_CTR_DRBG_USE_128_BIT_KEY
```
Comment (shelib01) [inline scripts/config.pl:59]: Why do we adding CTR_DRBG flag?

File: scripts/config.pl [hunk 5]
```diff
@@ -34,7 +36,7 @@
 #   MBEDTLS_REMOVE_3DES_CIPHERSUITES
 #   MBEDTLS_SHA256_NO_SHA224
 #   MBEDTLS_SSL_HW_RECORD_ACCEL
-#   MBEDTLS_SSL_PROTO_NO_DTLS
+#   MBEDTLS_SSL_PROTO_NO_TLS
```
Comment (shelib01) [inline scripts/config.pl:39]: What does this flag mean ?

File: scripts/config.pl [hunk 6]
```diff
@@ -98,9 +105,9 @@
 my @excluded = qw(
 MBEDTLS_TEST_NULL_ENTROPY
 MBEDTLS_DEPRECATED_REMOVED
+MBEDTLS_DEPRECATED_WARNING
 MBEDTLS_HAVE_SSE2
 MBEDTLS_PLATFORM_NO_STD_FUNCTIONS
-MBEDTLS_CTR_DRBG_USE_128_BIT_KEY
```
Comment (shelib01) [inline scripts/config.pl:103]: Why do we delete it from excluded flags?

File: scripts/config.pl [hunk 7]
```diff
@@ -109,14 +116,14 @@
 MBEDTLS_MEMORY_BUFFER_ALLOC_C
 MBEDTLS_NO_DEFAULT_ENTROPY_SOURCES
 MBEDTLS_NO_PLATFORM_ENTROPY
-MBEDTLS_RSA_NO_CRT
```
Comment (shelib01) [inline scripts/config.pl:112]: Why did we remove it from excluded flags ?

File: scripts/config.pl [hunk 8]
```diff
@@ -129,34 +136,38 @@
...
 
-# Things that should be enabled in "full" even if they match @excluded
+# Things that should be enabled in "full" even if they match @excluded.
+# Platform ALTs enable global variables that allow configuring the behavior
+# but default to the default behavior, except for PLATFORM_SETUP_TEARDOWN_ALT
+# which requires the application to provide relevant functions like
+# non-platform ALTs.
 my @non_excluded = qw(
-PLATFORM_[A-Z0-9]+_ALT
+PLATFORM_(?!SETUP_TEARDOWN_)[A-Z_0-9]+_ALT
```
Comment (shelib01) [inline scripts/config.pl:170]: Do you know what are they talking about ?

File: library/ssl_tls.c [hunk 9]
```diff
@@ -5263,16 +5277,22 @@ static int ssl_handle_possible_reconnect( mbedtls_ssl_context *ssl )
...
+        MBEDTLS_SSL_DEBUG_BUF( 4, "output record sent to network",
+                                  ssl->out_buf, len );
         /* Don't check write errors as we can't do anything here.
          * If the error is permanent we'll catch it later,
          * if it's not, then hopefully it'll work next time. */
-        (void) mbedtls_ssl_get_send( ssl )( ssl->p_bio, ssl->out_buf, len );
-        ret = 0;
+        send_ret = mbedtls_ssl_get_send( ssl )( ssl->p_bio, ssl->out_buf, len );
+        MBEDTLS_SSL_DEBUG_RET( 2, "mbedtls_ssl_get_send", send_ret );
+        (void) send_ret;
```
Comment (shelib01) [inline library/ssl_tls.c:5289]: Not sure I understand this return code implementation

**Comment by shelib01** [comment]
I didn't find changes in hmac_drbg.h as specified in comment

File: library/ssl_tls.c (hunk 1 above)
Comment (AndrzejKurek) [reply library/ssl_tls.c:1256]: This function modifies the context only if this flag is defined.

File: library/ssl_tls.c (hunk 2 above)
Comment (AndrzejKurek) [reply library/ssl_tls.c:5241]: The behavior was made different for the baremetal branch, and a fix was submitted for 2.16 and then ported in https://github.com/ARMmbed/mbedtls/pull/3149 to baremetal. It is now incorporated in this PR.

File: scripts/assemble_changelog.py (hunk 3 above)
Comment (AndrzejKurek) [reply scripts/assemble_changelog.py:4]: This is a new feature of mainstream Mbed TLS.

File: scripts/config.pl (hunk 4 above)
Comment (AndrzejKurek) [reply scripts/config.pl:59]: These flags were just reordered and cleaned up.

File: ChangeLog [hunk 10]
```diff
@@ -63,7 +63,24 @@ Changes
    * Reduce RAM consumption during session renegotiation by not storing
      the peer CRT chain and session ticket twice.
 
-= mbed TLS 2.16.X branch released XXXX-XX-XX
+= mbed TLS 2.16.6 branch released 2020-04-14
+
```
Comment (piotr-now) [inline ChangeLog:67]: Changes related with mbed TLS 2.16.6 are newer then changes described in 'mbed TLS "baremetal"' section. Shouldn't be added at the top of this file? Is it a convention that only baremetal changes are always at the top of the ChangeLog file?

File: tests/suites/test_suite_x509parse.data [hunk 11]
```diff
@@ -1380,6 +1380,14 @@ X509 CRT ASN1 (TBS, inv Validity, data remaining after 'notAfter')
 depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C
 x509parse_crt:"308198308182a0030201008204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301e170c303930313031303030303030170c3039313233313233353935391700300c310a30080600130454657374302a300d06092A864886F70D010101050003190030160210ffffffffffffffffffffffffffffffff0202ffff300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_X509_INVALID_DATE + MBEDTLS_ERR_ASN1_LENGTH_MISMATCH
 
+X509 Certificate ASN1 (inv extBasicConstraint, pathlen is INT_MAX)
```
Comment (piotr-now) [inline tests/suites/test_suite_x509parse.data:1383]:
```suggestion
X509 CRT ASN1 (inv extBasicConstraint, pathlen is INT_MAX)
```
The convention has changed from "X509 Certificate" to "X509 CRT"

File: tests/suites/test_suite_x509parse.data [hunk 12]
```diff
@@ -1380,6 +1380,14 @@ X509 CRT ASN1 (TBS, inv Validity, data remaining after 'notAfter')
...
+depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C:MBEDTLS_SHA1_C
+x509parse_crt_file:"data_files/server1_pathlen_int_max.crt":MBEDTLS_ERR_X509_INVALID_EXTENSIONS + MBEDTLS_ERR_ASN1_INVALID_LENGTH
+
+X509 Certificate ASN1 (pathlen is INT_MAX-1)
```
Comment (piotr-now) [inline tests/suites/test_suite_x509parse.data:1387]: The same as above

File: ChangeLog (hunk 10 above)
Comment (AndrzejKurek) [reply ChangeLog:67]: Looking at how it was handled previously - yes. Previous releases are lower, and baremetal is on top of that.

**Comment by AndrzejKurek** [comment]
Shortened test names just as Piotr suggested.

File: scripts/config.pl (hunk 5 above)
Comment (AndrzejKurek) [reply scripts/config.pl:39]: This is just some clean up work in the comments.

**Pull request merged**